File: MaxMiner.py
# MaxMiner.py
# this is my part, implements the Max-Miner algorithm
import csv
import math
import time
from collections import defaultdict


# loading the dataset
# no header in this csv, each line is just a transaction
# stripping whitespace because some rows had extra spaces

def load_basket_data(filepath):
    basket_list = []
    with open(filepath, newline="") as f:
        reader = csv.reader(f)
        for row in reader:
            cleaned = [item.strip() for item in row if item.strip()]
            if cleaned:
                basket_list.append(cleaned)
    return basket_list


# get every unique item in the dataset

def get_all_items(basket_list):
    all_items = set()
    for transaction in basket_list:
        all_items.update(transaction)
    return sorted(all_items)


# check if an itemset meets the support threshold

def is_frequent(itemset, basket_list, min_sup_count):
    count = sum(1 for t in basket_list if itemset.issubset(set(t)))
    return count >= min_sup_count


def get_support_count(itemset, basket_list):
    return sum(1 for t in basket_list if itemset.issubset(set(t)))


def is_subset_of_any_maximal(candidate, maximal_itemsets):
    # if we already have a bigger set that covers this one, skip it
    return any(candidate.issubset(mx) for mx in maximal_itemsets)


# MAX-MINER
# each node in the search is a (head, tail) pair
#   head = items we've committed to
#   tail = items we could still add
# if head+tail is frequent we keep extending
# if not, we trim the tail to only items that actually work with head
# this way we skip a ton of candidates that apriori would have checked

def run_max_miner(basket_list, min_sup_count):
    all_items = get_all_items(basket_list)
    maximal_itemsets = []

    # one starting node per item, tail is everything after it alphabetically
    # doing it this way avoids checking the same combo twice
    initial_nodes = []
    for i, item in enumerate(all_items):
        head = frozenset([item])
        tail = frozenset(all_items[i + 1:])
        initial_nodes.append((head, tail))

    queue = initial_nodes

    while queue:
        next_queue = []

        for head, tail in queue:
            full_set = head | tail

            if not tail:
                # nothing left to add, head is as big as it gets
                if is_frequent(head, basket_list, min_sup_count):
                    if not is_subset_of_any_maximal(head, maximal_itemsets):
                        maximal_itemsets.append(head)
                continue

            full_frequent = is_frequent(full_set, basket_list, min_sup_count)

            if full_frequent:
                # whole thing is frequent, record it and keep extending head
                if not is_subset_of_any_maximal(full_set, maximal_itemsets):
                    maximal_itemsets.append(full_set)
                tail_list = sorted(tail)
                for i, item in enumerate(tail_list):
                    new_head = head | frozenset([item])
                    new_tail = frozenset(tail_list[i + 1:])
                    next_queue.append((new_head, new_tail))
            else:
                # head+tail isn't frequent so trim tail to only items that work with head
                surviving_tail = frozenset(
                    item for item in tail
                    if is_frequent(head | frozenset([item]), basket_list, min_sup_count)
                )
                if is_frequent(head, basket_list, min_sup_count):
                    # head is frequent and can't be extended fully so it's maximal
                    if not is_subset_of_any_maximal(head, maximal_itemsets):
                        maximal_itemsets.append(head)

                if surviving_tail:
                    tail_list = sorted(surviving_tail)
                    for i, item in enumerate(tail_list):
                        new_head = head | frozenset([item])
                        new_tail = frozenset(tail_list[i + 1:])
                        next_queue.append((new_head, new_tail))

        queue = next_queue
    # remove anything that ended up being a subset of a bigger maximal set
    cleaned = []
    for mx in maximal_itemsets:
        if not any(mx < other for other in maximal_itemsets):
            cleaned.append(mx)

    return cleaned


# writing results to file
def itemset_to_str(frozen):
    return "{" + ", ".join(sorted(frozen)) + "}"


def write_results(max_miner_itemsets, basket_list, total_records,
                  min_sup, min_sup_count, maxminer_time, output_path):

    with open(output_path, "w") as f:
        f.write("Max-Miner Results\n")
        f.write("Group: Charlie Livingstone, Tung Vu, Efe Ivagba\n")
        f.write("Dataset: Market_Basket_Optimisation.csv\n")
        f.write(f"Total transactions: {total_records}\n")
        f.write(f"min_sup: {min_sup} (count = {min_sup_count})\n")
        f.write(f"Time taken: {maxminer_time:.4f} seconds\n")
        f.write(f"Maximal frequent itemsets found: {len(max_miner_itemsets)}\n\n")

        for i, itemset in enumerate(sorted(max_miner_itemsets, key=lambda x: -len(x)), 1):
            cnt = get_support_count(itemset, basket_list)
            sup = round(cnt / total_records, 2)
            f.write(f"Maximal#{i}: {itemset_to_str(itemset)} - support: {sup}\n")


# main
def main():
    dataset_filepath = "Market_Basket_Optimisation.csv"
    output_path = "MaxMiner_Results.txt"

    print("\nMax-Miner Algorithm")
    print("Efe Ivagba - B00909028\n")

    while True:
        try:
            min_sup_input = float(input("enter min_sup (e.g. 0.05): ").strip())
            if not (0 < min_sup_input <= 1):
                print("has to be between 0 and 1, try again")
                continue
            break
        except ValueError:
            print("enter a decimal like 0.05")

    print(f"\nloading {dataset_filepath} ...")
    basket_list = load_basket_data(dataset_filepath)
    total_records = len(basket_list)
    print(f"transactions loaded: {total_records}")

    min_sup_count = math.ceil(min_sup_input * total_records)
    print(f"min_sup count: ceil({min_sup_input} x {total_records}) = {min_sup_count}")

    print("\nrunning Max-Miner ...")
    t0 = time.time()
    maxminer_result = run_max_miner(basket_list, min_sup_count)
    maxminer_time = time.time() - t0
    print(f"done - {len(maxminer_result)} maximal itemsets found in {maxminer_time:.4f}s")

    write_results(maxminer_result, basket_list, total_records,
                  min_sup_input, min_sup_count, maxminer_time, output_path)

    print(f"\nresults saved to {output_path}\n")


if __name__ == "__main__":
    main()

File: MaxMiner.java
import java.io.*;
import java.util.*;

class Transaction {
    List<String> items;

    public Transaction(List<String> items) {
        this.items = items;
    }
}


public class MaxMiner {

    static List<Transaction> transactions = new ArrayList<>();
    static int min_sup_count;
    static double min_conf;

    static Map<Set<String>, Integer> maximalItemsets = new LinkedHashMap<>();

    public static void main(String[] args) throws IOException {

        // LOAD DATASET
        File dataset = new File("Market_Basket_Optimisation.csv");
        BufferedReader br = new BufferedReader(new FileReader(dataset));
        String line;

        while ((line = br.readLine()) != null) {
            String[] parts = line.split(",");
            List<String> items = new ArrayList<>();

            for (String item : parts) {
                item = item.trim();
                if (!item.isEmpty())
                    items.add(item);
            }

            if (!items.isEmpty())
                transactions.add(new Transaction(items));
        }
        br.close();

        Scanner kb = new Scanner(System.in);

        System.out.print("Enter minimum support: ");
        double minSupFraction = kb.nextDouble();

        System.out.print("Enter minimum confidence: ");
        min_conf = kb.nextDouble();

        min_sup_count = (int) Math.ceil(minSupFraction * transactions.size());

        Map<String, Integer> supportCache = new HashMap<>();
        for (String item : getAllItems()) {
            int sup = getSupport(Collections.singleton(item));
            if (sup >= min_sup_count)
                supportCache.put(item, sup);
        }

        List<String> allItems = new ArrayList<>(supportCache.keySet());

        allItems.sort(Comparator.comparingInt(supportCache::get));

        //Track runtime
        long startTime = System.nanoTime();

        maxMiner(new HashSet<>(), allItems);
        
        removeNonMaximal();

        long endTime = System.nanoTime();
        double runtimeSeconds = (endTime - startTime) / 1e9;

        writeRules(minSupFraction, min_conf, runtimeSeconds);

        System.out.println("Rules written to MaxMiner_Rules.txt");
    }

    public static List<String> getAllItems() {
        Set<String> items = new HashSet<>();
        for (Transaction t : transactions)
            items.addAll(t.items);
        return new ArrayList<>(items);
    }

    public static int getSupport(Set<String> itemset) { 

        int count = 0;
        for (Transaction t : transactions) {
            if (t.items.containsAll(itemset))
                count++;
        }
        return count;
    }

    // MAX-MINER Implementation based on Bayardo (1998) Figures 2–4
    // prefix = h(g)  (the head of the current candidate group)
    // tail   = t(g)  (ordered list of remaining items that may extend the head)
    public static void maxMiner(Set<String> prefix, List<String> tail) {

        // Superset-frequency pruning (Figure 2 / Figure 4)
        // If h ∪ t is frequent, every sub-node would produce a non-maximal itemset,
        // so record h ∪ t as a maximal candidate and stop expanding.
        Set<String> combined = new HashSet<>(prefix);
        combined.addAll(tail);

        if (!combined.isEmpty() && getSupport(combined) >= min_sup_count) {
            maximalItemsets.put(new HashSet<>(combined), getSupport(combined));
            return;
        }

        // Subset-infrequency pruning + tail building (Figure 4: GEN-SUB-NODES)
        // Prune tail items whose addition to the head makes the head infrequent,
        // then generate one sub-node per remaining tail item.
        for (int i = 0; i < tail.size(); i++) {

            String item = tail.get(i);

            Set<String> newPrefix = new HashSet<>(prefix);
            newPrefix.add(item);

            // Subset-infrequency pruning: skip item if h ∪ {item} is infrequent
            if (getSupport(newPrefix) < min_sup_count)
                continue;

            // Build the new tail: items after position i that are still potentially useful.
            // Per GEN-SUB-NODES we also prune tail items j where h' ∪ {j} is infrequent.
            List<String> newTail = new ArrayList<>();
            for (int j = i + 1; j < tail.size(); j++) {
                String tailItem = tail.get(j);
                Set<String> test = new HashSet<>(newPrefix);
                test.add(tailItem);
                if (getSupport(test) >= min_sup_count)
                    newTail.add(tailItem);
            }

            maxMiner(newPrefix, newTail);
        }

        // If prefix itself is frequent and nothing extended it, it may be maximal.
        // Record it here, the removeNonMaximal pass will clean up duplicates.
        if (!prefix.isEmpty() && getSupport(prefix) >= min_sup_count) {
            maximalItemsets.putIfAbsent(new HashSet<>(prefix), getSupport(prefix));
        }
    }

    // REMOVE NON-MAXIMAL
    // After the search, ensure every stored itemset has no proper frequent superset
    // also stored (per the "remove from F" step in Figure 2).
    public static void removeNonMaximal() {
        List<Set<String>> all = new ArrayList<>(maximalItemsets.keySet());
        Set<Set<String>> toRemove = new HashSet<>();

        for (int i = 0; i < all.size(); i++) {
            if (toRemove.contains(all.get(i))) continue;
            for (int j = 0; j < all.size(); j++) {
                if (i == j || toRemove.contains(all.get(j))) continue;
                // If all.get(j) is a proper superset of all.get(i), then all.get(i) is not maximal
                if (all.get(j).containsAll(all.get(i)) && all.get(j).size() > all.get(i).size()) {
                    toRemove.add(all.get(i));
                    break;
                }
            }
        }

        for (Set<String> s : toRemove)
            maximalItemsets.remove(s);
    }

    public static List<Set<String>> getSubsets(Set<String> set) {
        List<Set<String>> subsets = new ArrayList<>();
        List<String> list = new ArrayList<>(set);
        int n = list.size();
        int total = 1 << n;

        for (int i = 1; i < total - 1; i++) {
            Set<String> subset = new HashSet<>();
            for (int j = 0; j < n; j++) {
                if ((i & (1 << j)) > 0)
                    subset.add(list.get(j));
            }
            subsets.add(subset);
        }

        return subsets;
    }

    public static void writeRules(double minSupFraction, double minConf, double runtimeSeconds) throws IOException {

        PrintWriter pw = new PrintWriter(new FileWriter("MaxMiner_Rules.txt"));

        int totalTransactions = transactions.size();
        int ruleNum = 1;

        pw.println("1. User Input:");
        pw.println();
        pw.println("Support=" + minSupFraction);
        pw.println("Confidence=" + minConf);
        pw.println();
        pw.println("2. Maximal Frequent Itemsets Found: " + maximalItemsets.size());
        pw.println();
        pw.println("Runtime: " + runtimeSeconds + " seconds");
        System.out.println("Dataset: Market_Basket_Optimisation.csv");
        pw.println();
        pw.println("3. Rules:");
        pw.println();
        

        for (Set<String> itemset : maximalItemsets.keySet()) {

            if (itemset.size() < 2)
                continue;

            int itemsetSupport = maximalItemsets.get(itemset);

            List<Set<String>> subsets = getSubsets(itemset);

            for (Set<String> antecedent : subsets) {

                Set<String> consequent = new HashSet<>(itemset);
                consequent.removeAll(antecedent);

                if (consequent.isEmpty())
                    continue;

                int antecedentSupport = getSupport(antecedent);

                if (antecedentSupport == 0)
                    continue;

                double confidence = (double) itemsetSupport / antecedentSupport;
                double support = (double) itemsetSupport / totalTransactions;

                if (confidence >= minConf) {
                    pw.println("Rule#" + ruleNum + ": " + formatSet(antecedent) + " => " + formatSet(consequent));
                    pw.println("(Support=" + round(support) + ", Confidence=" + round(confidence) + ")");
                    pw.println();
                    ruleNum++;
                }
            }
        }

        pw.println("Total rules generated: " + (ruleNum - 1));
        pw.close();
    }

    public static String formatSet(Set<String> s) {
        StringBuilder sb = new StringBuilder("{");
        Iterator<String> it = s.iterator();
        while (it.hasNext()) {
            sb.append(it.next());
            if (it.hasNext()) sb.append(", ");
        }
        sb.append("}");
        return sb.toString();
    }

    public static double round(double value) {
        return Math.round(value * 100.0) / 100.0;
    }
}